src/pages/blog/2024-06-28-why-i-like-graphql.mdx

```diff
@@ -0,0 +1,92 @@
+query {
+  user(id: 321) {
+    handle # ✅ I am allowed to view Users public info
+    email # 🛑 I shouldn't be able to see their PII just because I can view the User
+  }
+  user(id: 123) {
+    blockedUsers {
+      # 🛑 And sometimes I shouldn't even be able to see their public info,
+      # because context matters!
+      handle
+    }
+  }
+}
+```
+
+`handle` and `email` both have different authorization rules. This is actually quite tricky to handle with a `GET /user/:id` endpoint as well. There's really nothing that makes GraphQL harder here. Yes when fine-grained authorization is needed, you'll need fine-grained authorization checks at that level.
+
+```graphql
+user(id: 123) {
+  blockedUsers {
+    # 🛑 And sometimes I shouldn't even be able to see their public info,
+    # because context matters!
+    handle
+  }
+}
+```
+
+This part is interesting as well and another challenge of authorizing code and models in general. The same "object" accessed through different contexts can actually have different authorization rules. Again, this is common in all API styles as well. That's why I actually usually advise designing this through a different model entirely, since it's likely they'll evolve differently. Here this is even possibly an API design mistake instead. If `handle` should never be seen for `blockedUsers`, then it shouldn't even be part of the schema here. This is a super common mistake where folks try to reuse common models/types instead of being specific.
+
+After 8 years of GraphQL, I realize more and more that authorization is much more than a GraphQL problem. The API layer part is easy, but most code-bases are much more complex and must guard against unauthorized access in much better ways. Companies like [oso](https://www.osohq.com/) and [authzed](https://authzed.com/) and two great examples of how to do this well but also how complex this thing can be in general.
+
+## Demand Control
+
+We cannot assume that all requests are equally hard on the server, whether we're using GraphQL or not.  The truth is that no matter what API style you use, whether that's a binary protocol over UDP or GraphQL, it is extremely rare, especially as the API surface grows, that all use-cases and "requests" will be equally expensive for a server to process.
+
+A very easy example to show this is simply a paginated endpoint:
+
+```
+GET /users?first=100
+
+GET /users?first=1
+```
+
+Or super expensive mutations:
+
+```
+POST /expensive-creation
+```
+
+To be 100% fair, of course GraphQL exposes this problem a bit more, earlier on, especially when not using persisted queries (Which should not happen!). And while folks building a small RPC API may not need to implement variable rate limiting or some sort of cost categories, they almost always end up having to.
+
+And again: this focuses on public, unauthenticated public APIs. I think we can agree this is not [GraphQL's Sweet Spot](https://productionreadygraphql.com/blog/2020-05-14-sweetspot).
```

Comment:
I would also add that unlike other APIs, with GraphQL, because of its static nature, you can and do have tools that help you with these calculations instead of only creating ad-hoc solutions per endpoint